### PR TITLE
Compile `ml-package-versions.yml` into a python module

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## Related Issues/PRs

 #7550

## What changes are proposed in this pull request?

This PR addes a script to compile `dev/update_ml_package_versions.py` into a python module containing a dictionary that represents `ml-package-versions` and resolves the issues #7550 raised. 

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/utils/autologging_utils/versioning.py [hunk 1]
```diff
@@ -57,29 +55,10 @@ def _strip_dev_version_suffix(version):
...
-
-        with as_file(files(mlflow).joinpath("ml-package-versions.yml")) as file:
-            version_file_path = file.as_posix()
-    else:
-        from importlib.resources import path
-
-        with path(mlflow, "ml-package-versions.yml") as file:
-            version_file_path = file.as_posix()
-    with open(version_file_path) as f:
-        return yaml.load(f, Loader=yaml.SafeLoader)
```
Comment (harupy) [inline mlflow/utils/autologging_utils/versioning.py:73]: On my machine, this function takes 26 ~ 27 milliseconds to finish.

File: dev/update_ml_package_versions.py [hunk 2]
```diff
@@ -89,6 +90,48 @@ def update_max_version(src, key, new_max_version, category):
...
+                "package_info": {
+                    "pip_release": pip_release,
+                },
+                "autologging": {
+                    "minimum": min_version,
+                    "maximum": max_version,
+                },
+            }
+
+        this_file = Path(__file__).resolve().relative_to(Path.cwd())
```
Comment (BenWilson2) [inline dev/update_ml_package_versions.py:123]: Will this give the correct resolution on nt-based file systems?
I seem to recall I had issues with pathing resolution in windows  for functionality in a different feature last year and it had some odd behavior when trying to access `__file__` attributes depending on where it was called from.
IIRC (I may be wrong here) using `from inspect import getsourcefile` and mapping over a no-op returns the current executable source location on all OS's. 

Would something like this work?

```
import inspect
from os.path import abspath

this_file = Path(abspath(inspect.getsourcefile(lambda:0))).resolve().relative_to(Path.cwd())
```

Comment (harupy) [reply dev/update_ml_package_versions.py:123]: We won't run this script on Windows·

Comment (BenWilson2) [reply dev/update_ml_package_versions.py:123]: sounds good :)

File: mlflow/ml_package_versions.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
+# This file is auto-generated by dev/compile_ml_package_versions.py. Please do not edit it manually.
```
Comment (harupy) [inline mlflow/ml_package_versions.py:1]: I'll add a check to ensure this file is up-to-date.

Comment (harupy) [reply mlflow/ml_package_versions.py:1]: Done.

**Review by WeichenXu123 — APPROVED** [review]

**Comment by WeichenXu123** [comment]
Shall we compile `pypi_package_index.json` as a python module as well ?

**Comment by harupy** [comment]
> Shall we compile pypi_package_index.json as a python module as well ?

Good question. I think we should but I won't in this PR.

File: pyproject.toml [hunk 4]
```diff
@@ -3,8 +3,9 @@ line-length = 100
 target-version = ['py37']
 force-exclude = '''
 /(
-  | mlflow/protos
-  | tests/protos
-  | examples/recipes
-)/
+    mlflow/protos/
```
Comment (BenWilson2) [inline pyproject.toml:6]: is the `|` not needed for the first element?

**Review by BenWilson2 — APPROVED** [review]
LGTM! Great idea Haru!

File: pyproject.toml (hunk 4 above)
Comment (harupy) [reply pyproject.toml:6]: No. `force-exclude` is a verbose regular expression. I think the right regex here is `(a|b|c)`, not `(|a|b|c)`.

Comment (harupy) [reply pyproject.toml:6]: `/` at line 5 represents the root.

**Pull request merged**